libintl/gettext.3
.\"	$Csoft$
.\"	$NetBSD: gettext.3,v 1.4 2000/10/31 16:11:32 itojun Exp $
.\"
.\" Copyright (c) 2000 Citrus Project,
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd October 20, 2000
.Dt GETTEXT 3
.Os
.Sh NAME
.Nm gettext ,
.Nm dgettext ,
.Nm ngettext ,
.Nm dngettext ,
.Nm textdomain ,
.Nm bindtextdomain ,
.Nm bind_textdomain_codeset ,
.Nm dcgettext ,
.Nm dcngettext
.Nd message handling functions
.Sh LIBRARY
.Lb libintl
.Sh SYNOPSIS
.Fd #include <libintl.h>
.Ft char *
.Fn gettext "const char *msgid"
.Ft char *
.Fn dgettext "const char *domainname" "const char *msgid"
.Ft char *
.Fn ngettext "const char *domainname" "const char *msgid" "unsigned long int n"
.Ft char *
.Fn dngettext "const char *domainname" "const char *msgid1" "const char *msgid2" "unsigned long int n"
.Ft char *
.Fn textdomain "const char *domainname"
.Ft char *
.Fn bindtextdomain "const char *domainname" "const char *dirname"
.Ft char *
.Fn bind_textdomain_codeset "const char *domainname" "const char *codeset"
.Fd #include <libintl.h>
.Fd #include <locale.h>
.Ft char *
.Fn dcgettext "const char *domainname" "const char *msgid" "int category"
.Ft char *
.Fn dcngettext "const char *domainname" "const char *msgid1" "const char *msgid2" "unsigned long int n" "int category"
.Sh DESCRIPTION
The
.Fn gettext ,
.Fn dgettext ,
and
.Fn dcgettext
functions attempt to retrieve a
target string based on the specified
.Fa msgid
argument within the context of a
specific domain and the current locale.
The length of strings returned by
.Fn gettext ,
.Fn dgettext ,
and
.fn dcgettext
is undetermined until the function is
called.
The
.Fa msgid
argument is a null-terminated string.
.Pp
The
.Fn ngettext ,
.Fn dngettext
and
.Fn dcngettext
functions are equivalent to
.Fn gettext ,
.Fn dgettext
and
.Fn dcgettext ,
respectively, except for the handling of
plural forms.
The
.Fn ngettext ,
.Fn dngettext
and
.Fn dcngettext
searches for the
message string using the
.Fa msgid1
argument as the key, using the argument
.Fa n
to
determine the plural form.
If no message catalogs are found,
.Fa msgid1
is returned
if
.Fa n Li == 1 ,
otherwise
.Fa msgid2
is returned.
.Pp
The
.Dv LANGUAGE
environment variable is examined first to determine the message
catalogs to be used.
The value of the
.Dv LANGUAGE
environment variable is a list
of locale names separated by colon (:) character.
If the
.Dv LANGUAGE
environment
variable is defined, each locale name is tried in the specified order and if a
message catalog containing the requested message is found, the message is
returned.
If the
.Dv LANGUAGE
environment variable is defined but failed to locate
a message catalog, the
.Fa msgid
string will be returned.
.Pp
If the
.Dv LANGUAGE
environment variable is not defined,
.Dv LC_ALL ,
.Dv LC_xxx
and
.Dv LANG
environment variables are examined to locate the message catalog, following the
convention used by the
.Xr setlocale 3
function.
.Pp
The pathname used to locate the message catalog is
.Pa dirname/locale/category/domainname.mo,
where dirname is the directory specified by 
.Fn bindtextdomain ,
locale is a locale name determined by the definition of environment variables,
.Fa category
is
.Dv LC_MESSAGES
if 
.Fn gettext ,
.Fn ngettext ,
.Fn dgettext
or 
.Fn dngettext
is
called, otherwise
.Dv LC_xxx
where the name is the same as the locale category name
specified by the
.Fa category
argument of 
.Fn dcgettext
or 
.Fn dcngettext .
.Fa domainname
is the name of the domain specified by 
.Fn textdomain
or the
.Fa domainname
argument of 
.Fn dgettext ,
.Fn dngettext ,
.Fn dcgettext
or 
.Fn dcngettext .
.Pp
For 
.Fn gettext
and 
.Fn ngettext ,
the domain used is set by the last valid call to
.Fn textdomain .
If a valid call to 
.Fn textdomain
has not been made, the default
domain (called messages) is used.
.Pp
For 
.Fn dgettext ,
.Fn dngettext ,
.Fn dcgettext
and 
.Fn dcngettext ,
the domain used is
specified by the
.Fa domainname
argument.
The
.Fa domainname
argument is equivalent in
syntax and meaning to the
.Fa domainname
argument to 
.Fn textdomain ,
except that the
selection of the domain is valid only for the duration of the 
.Fn dgettext ,
.Fn dngettext ,
.Fn dcgettext
or 
.Fn dcngettext
function call.
.Pp
The 
.Fn dcgettext
and 
.Fn dcngettext
functions require additional argument
.Fa category
for retrieving message string for other than
.Dv LC_MESSAGES
category.
Available value for the
.Fa category
argument are
.Dv LC_CTYPE ,
.Dv LC_COLLATE ,
.Dv LC_MESSAGES ,
.Dv LC_MONETARY ,
.Dv LC_NUMERIC
and
.Dv LC_TIME .
The call of
.Fn dcgettext "domainname" "msgid" "LC_MESSAGES"
is equivalent to
.Fn dgettext "domainname" "msgid" .
Note that
.Dv LC_ALL
must not be used.
.Pp
The
.Fn textdomain
function sets or queries the name of the current domain of the
active
.Dv LC_MESSAGES
locale category.
The
.Fa domainname
argument is a
null-terminated string that can contain only the characters allowed in legal
filenames.
.Pp
The
.Fa domainname
argument is the unique name of a domain on the system.
If there
are multiple versions of the same domain on one system, namespace collisions
can be avoided by using
.Fn bindtextdomain .
If
.Fn textdomain
is not called, a
default domain is selected.
The setting of domain made by the last valid call
to
.Fn textdomain
remains valid across subsequent calls to
.Xr setlocale 3 ,
and
.Fn gettext .
.Pp
The
.Fa domainname
argument is applied to the currently active LC_MESSAGES locale.
.Pp
The current setting of the domain can be queried without affecting the current
state of the domain by calling
.Fn textdomain
with
.Fa domainname
set to the null pointer.
Calling
.Fn textdomain
with a
.Fa domainname
argument of a null string sets
the domain to the default domain
.Pq messages .
.Pp
The
.Fn bindtextdomain
function binds the path predicate for a message domain
.Fa domainname
to the value contained in dirname.
If
.Fa domainname
is a non-empty
string and has not been bound previously,
.Fn bindtextdomain
binds
.Fa domainname
with
.Fa dirname .
.Pp
If
.Fa domainname
is a non-empty string and has been bound previously,
.Fn bindtextdomain
replaces the old binding with dirname.
The dirname argument
can be an absolute pathname being resolved when
.Fn gettext ,
.Fn ngettext ,
.Fn dgettext ,
.Fn dngettext ,
.Fn dcgettext ,
or
.Fn dcngettext
are called.
If
.Fa domainname
is a null pointer or an empty string,
.Fn bindtextdomain
returns null pointer.
If
.Fn bindtextdomain
is not called, implementation-defined default directory is used.
.Pp
The
.Fn bind_textdomain_codeset
function can be used to specify the output
.Fa codeset
for message catalogs for domain
.Fa domainname .
The
.Fa codeset
argument must
be a valid codeset name which can be used for the
.Xr iconv_open 3
function.
.Pp
If the
.Fa codeset
argument is the null pointer,
.Fn bind_textdomain_codeset
returns the currently selected
.Fa codeset
for the domain with the name
.Fa domainname .
It returns null pointer if no
.Fa codeset
has yet been selected.
.Pp
The
.Fn bind_textdomain_codeset
function can be used several times.
If used multiple times, with the same
.Fa domainname
argument,
the later call overrides the
settings made by the earlier one.
.Pp
The
.Fn bind_textdomain_codeset
function returns a pointer to a string containing
the name of the selected
.Fa codeset .
.\".Sh "RETURN VALUES"
.\".Sh EXAMPLES
.Sh SEE ALSO
.Xr setlocale 3
.\".Sh STANDARDS
.Sh HISTORY
The functions are implemented by Citrus project,
based on the documentations for GNU gettext.
.Sh BUGS
The text was ripped off from Annex C of
.Dq LI18NUX 2000 Globalization Specification Version 1.0 .
.Pp
.Fn bind_textdomain_codeset
does not work at this moment
.Pq always fail .


CHANGELOG.md
# Changelog
All notable changes to Agar will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.8.0] - Unreleased
### Added
- [**Style Editor**](https://libagar.org/man3/AG_StyleEditor): Display computed style rules in addition to per-instance overrides.
- [**AG_Button**](https://libagar.org/man3/AG_Button): New style attributes `text-align` and `vertical-align`.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Introduce the `style()` method, called by the style compiler to parse widget-specific style attributes into an efficient binary representation.
- [**AG_Object**](https://libagar.org/man3/AG_Object): New function `AG_AccessVariable_NoLock()` (lockless variant of `AG_AccessVariable()`).
- [**AG_Object**](https://libagar.org/man3/AG_Object): Document the `AG_OBJECT_BOUND_EVENTS` flag.

### Removed
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove the "window-detached" event. It is redundant since the generic "detached" event does the same job.

### Changed
- Major performance improvements in the style compiler. Since it runs whenever widgets are resized, this should lead to better fluidity and overall lower CPU usage when Agar windows and container widgets are resized.
- [**AG_Object**](https://libagar.org/man3/AG_Object): In `AG_ObjectDestroy()`, do not call `AG_ObjectDetach()` on child objects. Instead, just cancel any active timers and null their parent pointers without raising a "detached" event. This allows us to perform garbage collection more efficiently.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Make `AG_TLIST_NO_BGLINES` the default.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): In `AG_TlistNewPolledMs()`, accept a `ms` argument of 0 to disable periodical updates (the caller may still trigger updates with "tlist-poll").
- In `AG_WidgetCopyStyle()`, prevent redundant locking by using `AG_AccessVariable_NoLock()`.

### Fixed
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Fix redundant "detached" handler invocations. Improve overall efficiency of the garbage-collection step after windows are destroyed.
- [**AG_Console**](https://libagar.org/man3/AG_Console): Prevent a redundant lock operation in the backend of `AG_ConsoleMsg()`.

## [1.7.1] - 2025-11-15
### Added
- An awesome new Agar logo. Thanks [eightbit](https://8bitprodigy.github.io)!
- Added support for the [Cmake](https://cmake.org) build system.
- Added support for the [Meson](https://mesonbuild.com) build system. Thanks [coleman](https://git.sr.ht/~coleman/)!
- New functions [**AG_InitVideoSDL2**](https://libagar.org/man3/AG_InitVideoSDL2) and [**AG_SetVideoSurfaceSDL2**](https://libagar.org/man3/AG_SetVideoSurfaceSDL2) for integrating with an existing SDL2 display context. Thanks Brigham Keys!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Add "Any File" type.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New function `AG_TlistCopy()`. Copy all items from a source to a destination `AG_Tlist`.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New constructor `AG_ComboNewFn()`. Sets a `combo-expanded` handler.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New member `nVisItems`. Set the number of items to show by default in expansions.
- [**AG_Checkbox**](https://libagar.org/man3/AG_Checkbox): New functions `AG_CheckboxText()` and `AG_CheckboxTextS()` to update the text label.
- Tweak the default color scheme. Document style defaults as a comment block in `gui/style.css`.
- New Ada bindings for the `Window`, `Box`, `Button`, `Checkbox` and `Combo` widget classes. Updated `ada/gui/demo`.

### Removed
- [**AG_Db**](https://libagar.org/man3/AG_Db): Remove support for DB5.
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove an unnecessary counter variable in `SizeAllocate()`.

### Changed
- Small cosmetic changes to the default color scheme. Set default font size back to 14pt.
- [**AG_Titlebar**](https://libagar.org/man3/AG_Titlebar): Enable the `AG_BUTTON_EXCL` optimization on titlebar buttons.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Allow the dropdown list to be populated by a `combo-expanded` handler.
- [**AG_UCombo**](https://libagar.org/man3/AG_UCombo): Allow the dropdown list to be populated by a `ucombo-expanded` handler.

### Fixed
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Make it again possible to statically initialize `list` before `combo-expanded`. Restores compatibility pre-1.6. Thanks Wally!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix widget geometries not updating when switching to a different Type filter.
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix compilation where `GLOB_TILDE` is unavailable.
- SDL2 drivers: Require at least version 2.0.22 of SDL2 (`SDL_HINT_MOUSE_AUTO_CAPTURE` is necessary for best user experience).
- [**AG_ProgressBar**](https://libagar.org/man3/AG_ProgressBar): Make `padding` work as expected in progress bar. Thanks scaramacai!
- [**cocoa**](https://libagar.org/man3/AG_DriverCocoa): Fix a compilation error.
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Fix a keyboard handling error. Don't pass `KeyRelease` events to `XwcLookupString()`. Fixes issue #74. Thanks @tsesani and @scaramacai!
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Introduce the `no-im` (disable input methods) driver option.
- [**MAP**](https://libagar.org/man3/MAP): Fix a crash when attempting to create an object if `pLibsRoot` is NULL.
- [**AG_Web**](https://libagar.org/man3/AG_Web): Fix inline compilation problem due to a typo in `net/web.h`.
- On Darwin platforms, build with `_DARWIN_C_SOURCE` globally. Fixes issue #77. Thanks @kwhr0!
- On FreeBSD check for availability of `CLOCK_SECOND` in addition to `HAVE_CLOCK_GETTIME`.
- Test for the `dirfd()` function at compilation.

trunk/libintl/gettext.c
/*	$Csoft: gettext.c,v 1.2 2003/08/06 04:08:16 vedge Exp $	*/
/*	$NetBSD: gettext.c,v 1.9 2001/02/16 07:20:35 minoura Exp $	*/

/*-
 * Copyright (c) 2000, 2001 Citrus Project,
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <config/localedir.h>
#include <config/enable_nls.h>

#ifdef ENABLE_NLS

#include <sys/types.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include <sys/uio.h>

#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#if 0
#include <util.h>
#endif
#include "libintl.h"
#include <locale.h>
#include "libintl_local.h"

static const char *lookup_category(int);
static const char *split_locale(const char *);
static const char *lookup_mofile(char *, size_t, const char *,
	char *, const char *, const char *, struct domainbinding *);
static u_int32_t flip(u_int32_t, u_int32_t);
static int validate(void *, struct mohandle *);
static int mapit(const char *, struct domainbinding *);
static int unmapit(struct domainbinding *);
static const char *lookup_hash(const char *, struct domainbinding *);
static const char *lookup_bsearch(const char *, struct domainbinding *);
static const char *lookup(const char *, struct domainbinding *);

/*
 * shortcut functions.  the main implementation resides in dcngettext().
 */
char *
gettext(msgid)
	const char *msgid;
{

	return dcngettext(NULL, msgid, NULL, 1UL, LC_MESSAGES);
}

char *
dgettext(domainname, msgid)
	const char *domainname;
	const char *msgid;
{

	return dcngettext(domainname, msgid, NULL, 1UL, LC_MESSAGES);
}

char *
dcgettext(domainname, msgid, category)
	const char *domainname;
	const char *msgid;
	int category;
{

	return dcngettext(domainname, msgid, NULL, 1UL, category);
}

char *
ngettext(msgid1, msgid2, n)
	const char *msgid1;
	const char *msgid2;
	unsigned long int n;
{

	return dcngettext(NULL, msgid1, msgid2, n, LC_MESSAGES);
}

char *
dngettext(domainname, msgid1, msgid2, n)
	const char *domainname;
	const char *msgid1;
	const char *msgid2;
	unsigned long int n;
{

	return dcngettext(domainname, msgid1, msgid2, n, LC_MESSAGES);
}

/*
 * dcngettext() -
 * lookup internationalized message on database locale/category/domainname
 * (like ja_JP.eucJP/LC_MESSAGES/domainname).
 * if n equals to 1, internationalized message will be looked up for msgid1.
 * otherwise, message will be looked up for msgid2.
 * if the lookup fails, the function will return msgid1 or msgid2 as is.
 *
 * Even though the return type is "char *", caller should not rewrite the
 * region pointed to by the return value (should be "const char *", but can't
 * change it for compatibility with other implementations).
 *
 * by default (if domainname == NULL), domainname is taken from the value set
 * by textdomain().  usually name of the application (like "ls") is used as
 * domainname.  category is usually LC_MESSAGES.
 *
 * the code reads in *.mo files generated by GNU gettext.  *.mo is a host-
 * endian encoded file.  both endians are supported here, as the files are in
 * /usr/share/locale! (or we should move those files into /usr/libdata)
 */

static const char *
lookup_category(category)
	int category;
{

	switch (category) {
	case LC_COLLATE:	return "LC_COLLATE";
	case LC_CTYPE:		return "LC_CTYPE";
	case LC_MONETARY:	return "LC_MONETARY";
	case LC_NUMERIC:	return "LC_NUMERIC";
	case LC_TIME:		return "LC_TIME";
	case LC_MESSAGES:	return "LC_MESSAGES";
	}
	return NULL;
}

/*
 * XPG syntax: language[_territory[.codeset]][@modifier]
 * XXX boundary check on "result" is lacking
 */
static const char *
split_locale(lname)
	const char *lname;
{
	char buf[BUFSIZ], tmp[BUFSIZ];
	char *l, *t, *c, *m;
	static char result[BUFSIZ];

	memset(result, 0, sizeof(result));

	if (strlen(lname) + 1 > sizeof(buf)) {
fail:
		return lname;
	}

	Strlcpy(buf, lname, sizeof(buf));
	m = strrchr(buf, '@');
	if (m)
		*m++ = '\0';
	c = strrchr(buf, '.');
	if (c)
		*c++ = '\0';
	t = strrchr(buf, '_');
	if (t)
		*t++ = '\0';
	l = buf;
	if (strlen(l) == 0)
		goto fail;
	if (c && !t)
		goto fail;

	if (m) {
		if (t) {
			if (c) {
				snprintf(tmp, sizeof(tmp), "%s_%s.%s@%s",
				    l, t, c, m); 
				Strlcat(result, tmp, sizeof(result));
				Strlcat(result, ":", sizeof(result));
			}
			snprintf(tmp, sizeof(tmp), "%s_%s@%s", l, t, m); 
			Strlcat(result, tmp, sizeof(result));
			Strlcat(result, ":", sizeof(result));
		}
		snprintf(tmp, sizeof(tmp), "%s@%s", l, m); 
		Strlcat(result, tmp, sizeof(result));
		Strlcat(result, ":", sizeof(result));
	}
	if (t) {
		if (c) {
			snprintf(tmp, sizeof(tmp), "%s_%s.%s", l, t, c); 
			Strlcat(result, tmp, sizeof(result));
			Strlcat(result, ":", sizeof(result));
		}
		Strlcat(result, tmp, sizeof(result));
		Strlcat(result, ":", sizeof(result));
	}
	Strlcat(result, l, sizeof(result));

	return result;
}

static const char *
lookup_mofile(buf, len, dir, lpath, category, domainname, db)
	char *buf;
	size_t len;
	const char *dir;
	char *lpath;	/* list of locales to be tried */
	const char *category;
	const char *domainname;
	struct domainbinding *db;
{
	struct stat st;
	char *p, *q;

	q = lpath;
	/* CONSTCOND */
	while (1) {
		p = strsep(&q, ":");
		if (!p)
			break;
		if (!*p)
			continue;

		/* don't mess with default locales */
		if (strcmp(p, "C") == 0 || strcmp(p, "POSIX") == 0)
			return NULL;

		/* validate pathname */
		if (strchr(p, '/') || strchr(category, '/'))
			continue;
#if 1	/*?*/
		if (strchr(domainname, '/'))
			continue;
#endif

		Strlcpy(buf, dir, len);
		Strlcat(buf, "/", len);
		Strlcat(buf, p, len);
		Strlcat(buf, "/", len);
		Strlcat(buf, category, len);
		Strlcat(buf, "/", len);
		Strlcat(buf, domainname, len);
		Strlcat(buf, ".mo", len);

		if (stat(buf, &st) < 0)
			continue;
		if ((st.st_mode & S_IFMT) != S_IFREG)
			continue;

		if (mapit(buf, db) == 0)
			return buf;
	}

	return NULL;
}

static u_int32_t
flip(v, magic)
	u_int32_t v;
	u_int32_t magic;
{

	if (magic == MO_MAGIC)
		return v;
	else if (magic == MO_MAGIC_SWAPPED) {
		v = ((v >> 24) & 0xff) | ((v >> 8) & 0xff00) |
		    ((v << 8) & 0xff0000) | ((v << 24) & 0xff000000);
		return v;
	} else {
		abort();
		/*NOTREACHED*/
	}
}

static int
validate(arg, mohandle)
	void *arg;
	struct mohandle *mohandle;
{
	char *p;

	p = (char *)arg;
	if (p < (char *)mohandle->addr ||
	    p > (char *)mohandle->addr + mohandle->len)
		return 0;
	else
		return 1;
}

int
mapit(path, db)
	const char *path;
	struct domainbinding *db;
{
	int fd;
	struct stat st;
	char *base;
	u_int32_t magic, revision;
	struct moentry *otable, *ttable;
	struct moentry_h *p;
	struct mo *mo;
	size_t l;
	int i;
	char *v;
	struct mohandle *mohandle = &db->mohandle;

	if (mohandle->addr && mohandle->addr != MAP_FAILED &&
	    mohandle->mo.mo_magic)
		return 0;	/*already opened*/

	unmapit(db);

#if 0
	if (secure_path(path) != 0)
		goto fail;
#endif
	if (stat(path, &st) < 0)
		goto fail;
	if ((st.st_mode & S_IFMT) != S_IFREG || st.st_size > GETTEXT_MMAP_MAX)
		goto fail;
	fd = open(path, O_RDONLY);
	if (fd < 0)
		goto fail;
	if (read(fd, &magic, sizeof(magic)) != sizeof(magic) ||
	    (magic != MO_MAGIC && magic != MO_MAGIC_SWAPPED)) {
		close(fd);
		goto fail;
	}
	if (read(fd, &revision, sizeof(revision)) != sizeof(revision) ||
	    flip(revision, magic) != MO_REVISION) {
		close(fd);
		goto fail;
	}
	mohandle->addr = mmap(NULL, (size_t)st.st_size, PROT_READ,
	    MAP_FILE | MAP_SHARED, fd, (off_t)0);
	if (!mohandle->addr || mohandle->addr == MAP_FAILED) {
		close(fd);
		goto fail;
	}
	close(fd);
	mohandle->len = (size_t)st.st_size;

	base = mohandle->addr;
	mo = (struct mo *)mohandle->addr;

	/* flip endian.  do not flip magic number! */
	mohandle->mo.mo_magic = mo->mo_magic;
	mohandle->mo.mo_revision = flip(mo->mo_revision, magic);
	mohandle->mo.mo_nstring = flip(mo->mo_nstring, magic);

	/* validate otable/ttable */
	otable = (struct moentry *)(base + flip(mo->mo_otable, magic));
	ttable = (struct moentry *)(base + flip(mo->mo_ttable, magic));
	if (!validate(otable, mohandle) ||
	    !validate(&otable[mohandle->mo.mo_nstring], mohandle)) {
		unmapit(db);
		goto fail;
	}
	if (!validate(ttable, mohandle) ||
	    !validate(&ttable[mohandle->mo.mo_nstring], mohandle)) {
		unmapit(db);
		goto fail;
	}

	/* allocate [ot]table, and convert to normal pointer representation. */
	l = sizeof(struct moentry_h) * mohandle->mo.mo_nstring;
	mohandle->mo.mo_otable = (struct moentry_h *)malloc(l);
	if (!mohandle->mo.mo_otable) {
		unmapit(db);
		goto fail;
	}
	mohandle->mo.mo_ttable = (struct moentry_h *)malloc(l);
	if (!mohandle->mo.mo_ttable) {
		unmapit(db);
		goto fail;
	}
	p = mohandle->mo.mo_otable;
	for (i = 0; i < mohandle->mo.mo_nstring; i++) {
		p[i].len = flip(otable[i].len, magic);
		p[i].off = base + flip(otable[i].off, magic);

		if (!validate(p[i].off, mohandle) ||
		    !validate(p[i].off + p[i].len + 1, mohandle)) {
			unmapit(db);
			goto fail;
		}
	}
	p = mohandle->mo.mo_ttable;
	for (i = 0; i < mohandle->mo.mo_nstring; i++) {
		p[i].len = flip(ttable[i].len, magic);
		p[i].off = base + flip(ttable[i].off, magic);

		if (!validate(p[i].off, mohandle) ||
		    !validate(p[i].off + p[i].len + 1, mohandle)) {
			unmapit(db);
			goto fail;
		}
	}

	/* grab MIME-header and charset field */
	mohandle->mo.mo_header = lookup("", db);
	if (mohandle->mo.mo_header)
		v = strstr(mohandle->mo.mo_header, "charset=");
	else
		v = NULL;
	if (v) {
		mohandle->mo.mo_charset = strdup(v + 8);
		if (!mohandle->mo.mo_charset)
			goto fail;
		v = strchr(mohandle->mo.mo_charset, '\n');
		if (v)
			*v = '\0';
	}

	/*
	 * XXX check charset, reject it if we are unable to support the charset
	 * with the current locale.
	 * for example, if we are using euc-jp locale and we are looking at
	 * *.mo file encoded by euc-kr (charset=euc-kr), we should reject
	 * the *.mo file as we cannot support it.
	 */

	return 0;

fail:
	return -1;
}

static int
unmapit(db)
	struct domainbinding *db;
{
	struct mohandle *mohandle = &db->mohandle;

	/* unmap if there's already mapped region */
	if (mohandle->addr && mohandle->addr != MAP_FAILED)
		munmap(mohandle->addr, mohandle->len);
	mohandle->addr = NULL;
	if (mohandle->mo.mo_otable)
		free(mohandle->mo.mo_otable);
	if (mohandle->mo.mo_ttable)
		free(mohandle->mo.mo_ttable);
	if (mohandle->mo.mo_charset)
		free(mohandle->mo.mo_charset);
	memset(&mohandle->mo, 0, sizeof(mohandle->mo));
	return 0;
}

/* ARGSUSED */
static const char *
lookup_hash(msgid, db)
	const char *msgid;
	struct domainbinding *db;
{

	/*
	 * XXX should try a hashed lookup here, but to do so, we need to
	 * look inside the GPL'ed *.c and re-implement...
	 */
	return NULL;
}

static const char *
lookup_bsearch(msgid, db)
	const char *msgid;
	struct domainbinding *db;
{
	int top, bottom, middle, omiddle;
	int n;
	struct mohandle *mohandle = &db->mohandle;

	top = 0;
	bottom = mohandle->mo.mo_nstring;
	omiddle = -1;
	/* CONSTCOND */
	while (1) {
		if (top > bottom)
			break;
		middle = (top + bottom) / 2;
		/* avoid possible infinite loop, when the data is not sorted */
		if (omiddle == middle)
			break;
		if (middle < 0 || middle >= mohandle->mo.mo_nstring)
			break;

		n = strcmp(msgid, mohandle->mo.mo_otable[middle].off);
		if (n == 0)
			return (const char *)mohandle->mo.mo_ttable[middle].off;
		else if (n < 0)
			bottom = middle;
		else
			top = middle;
		omiddle = middle;
	}

	return NULL;
}

static const char *
lookup(msgid, db)
	const char *msgid;
	struct domainbinding *db;
{
	const char *v;

	v = lookup_hash(msgid, db);
	if (v)
		return v;

	return lookup_bsearch(msgid, db);
}

char *
dcngettext(domainname, msgid1, msgid2, n, category)
	const char *domainname;
	const char *msgid1;
	const char *msgid2;
	unsigned long int n;
	int category;
{
	const char *msgid;
	char path[PATH_MAX];
	static char lpath[PATH_MAX];
	static char olpath[PATH_MAX];
	const char *locale;
	const char *language;
	const char *cname = NULL;
	const char *v;
	static char *ocname = NULL;
	static char *odomainname = NULL;
	struct domainbinding *db;

	msgid = (n == 1) ? msgid1 : msgid2;
	if (msgid == NULL)
		return NULL;

	if (!domainname)
		domainname = __current_domainname;
	cname = lookup_category(category);
	if (!domainname || !cname) {
		goto fail;
	}

	if ((language = getenv("LANGUAGE")) == NULL) {
		language = getenv("LANG");
	}

	locale = setlocale(LC_MESSAGES, NULL);	/*XXX*/
	if (locale)
		locale = split_locale(locale);
	if (language && locale) {
		if (strlen(language) + strlen(locale) + 2 > sizeof(lpath))
			goto fail;
		Strlcpy(lpath, language, sizeof(lpath));
		Strlcat(lpath, ":", sizeof(lpath));
		Strlcat(lpath, locale, sizeof(lpath));
	} else if (language) {
		if (strlen(language) + 1 > sizeof(lpath))
			goto fail;
		Strlcpy(lpath, language, sizeof(lpath));
	} else if (locale) {
		if (strlen(locale) + 1 > sizeof(lpath))
			goto fail;
		Strlcpy(lpath, locale, sizeof(lpath));
	} else
		goto fail;

	for (db = __bindings; db; db = db->next)
		if (strcmp(db->domainname, domainname) == 0)
			break;
	if (!db) {
		if (!bindtextdomain(domainname, LOCALEDIR))
			goto fail;
		db = __bindings;
	}

	/* don't bother looking it up if the values are the same */
	if (odomainname && strcmp(domainname, odomainname) == 0 &&
	    ocname && strcmp(cname, ocname) == 0 &&
	    strcmp(lpath, olpath) == 0 &&
	    db->mohandle.mo.mo_magic)
		goto found;

	/* try to find appropriate file, from $LANGUAGE */
	if (lookup_mofile(path, sizeof(path), db->path, lpath, cname,
	    domainname, db) == NULL)
		goto fail;

	if (odomainname)
		free(odomainname);
	if (ocname)
		free(ocname);
	odomainname = strdup(domainname);
	ocname = strdup(cname);
	if (!odomainname || !ocname) {
		if (odomainname)
			free(odomainname);
		if (ocname)
			free(ocname);
		odomainname = ocname = NULL;
		goto fail;
	}

	Strlcpy(olpath, lpath, sizeof(olpath));

found:
	v = lookup(msgid, db);
	if (v) {
		/*
		 * XXX call iconv() here, if translated text is encoded
		 * differently from currently-selected encoding (locale).
		 * look at Content-type header in *.mo file, in string obtained
		 * by gettext("").
		 */

		/*
		 * Given the amount of printf-format security issues, it may
		 * be a good idea to validate if the original msgid and the
		 * translated message format string carry the same printf-like
		 * format identifiers.
		 */

		msgid = v;
	}

fail:
	/* LINTED const cast */
	return (char *)msgid;
}

#endif	/* ENABLE_NLS */
